### ISSUE TITLE
Rule Idea: code instead of variable/mixin use

### ISSUE THREAD

**Issue by branpar** [kristerkari/stylelint-scss#80]
I have a tough but helpful rule(s) to add. They're similar but should be split into two rules.

**Rule 1:**
Let's say I have a variable `$color-gray90: #111;` . Then, in my code I see `color: #111;` I would want a rule to check for this and alert that we should be using the variable instead of the declaration.

**Rule 2:** 
Similar but perhaps more important, let's say I have defined a mixin:

```
@mixin opacity($opacity) {
  opacity: $opacity;
  $opacity-ie: $opacity * 100;
  filter: alpha(opacity=$opacity-ie);
}
```

Then, in my code I see the following:

```
button {
  opacity: 0.8;
  filter: alpha(opacity=80);
}
```

I would like a rule to identify that the user should have used the mixin instead of the definitions directly. This becomes even more valuable when: 1) linting legacy or older code, or 2) when linting after creating a new mixin to determine where else it can be placed.

I find great value in both of these rules. Could they be added to the list of potential additions? They're probably tough to make but would be extremely helpful.

**Comment by dryoma** [comment]
Hi @branpar thanks for you suggestions.

1) For this we already have a proposal, even two: #4 and #13. Still to be implemented though.
2) A valid point either, should give it a thought. Meanwhile, you could utilize stylelint's [`property-blacklist`](http://stylelint.io/user-guide/rules/property-blacklist/) and/or [`declaration-property-value-blacklist`](http://stylelint.io/user-guide/rules/declaration-property-value-blacklist/). In fact, the rule that you proposed kind of overlaps with them. I.e. `declaration-property-value-blacklist` can prohibit certain properties and value patterns, and the proposed rule does the same, the only difference would be a warning message. Do you think the existing rules will do?

Anyway, any thoughts on the 2 guys?

**Comment by branpar** [comment]
I can see the idea you're driving towards with the blacklists. The problem is that if you have many mixins you have to have a substantial blacklist, and the mixins themselves will violate the rules which means then you have to add in exception comments. Additionally each time someone adds a mixin you have to add to the blacklist as well. This doesn't seem like a realistic workflow on a large team, like mine.

It would be better for it to read the mixins based off the @mixin keyword and then match rules based on these in the code. In my mixin example above it should match _any_ property within the mixin. So if a selector only had `opacity: 0.8`, the rule would trip because we want to use the mixin so that we also include vendor prefixes and IE support.

**Comment by dryoma** [comment]
I see. Though there are other issues with such approach.

The main issue I see is that stylelint parses only one file at a time, thus the linter would warn only on the properties that are used in mixins inside the same file. I.e. if that `@mixn opacity` is in `_mixins.scss`, the linter won't give a warning on an a declaration with `opacity` in, say, `_header.scss` unless it has a similar (or the same) mixin defined in it.

2) It would also be troublesome if a user would need some mixins with more common props like `display`, `font-size`, i.e. that are not intended to be such "must replace" helpers that you described. So the rule would have to be turned off for every such declaraton inside rulesets. E.g.:

```scss
@mixin vert-centered ($n) {
  position: absolute;
  top: 50%;
  transform: translate(0, -50%);
}

// position, top and transform here should be replaced by a mixin call
.play-btn {
  position: absolute;
  top: 50%;
  font: 20px/1 $icon-font;
  transform: translate(0, -50%);
}

.video-container {
  // false positive; we'll have to turn off the rule for this decl (and any other decl like this).
  position: relative;
  ...
}
```

**Comment by branpar** [comment]
Good and valid points. Assuming that users organize their mixins in a _mixin.scss file perhaps the rule could take in an array of files that contain mixins so that it could pre-parse those? Variables could have a similar solution but I suspect variables are more widespread among files. But in order to fix that there would have to be some sort of discovery pre-parsing to determine things like mixin and variable declarations in order to be able to check against them. 

For your second point I don't have a ready solution for that. It's a really good point. So sounds like this would really only work for a 1-to-1 match among the items in the mixin vs what's in the selector block. If there's an exact match it would fail the rule, otherwise it wouldn't. Not ideal to say the least when mixins get complex.

**Comment by dryoma** [comment]
In terms of variables it's both simpler and more complex, actually. The simpler part is that we have to *just* find where a certain type of a value is and if it's not a variable. The complex part is that *just*: the parsing.

As for mixins,

> there would have to be some sort of discovery pre-parsing

Considering such pre-parsing would have to be done for *every file* that needs to be linted, the more files there are the longer it will get. @kristerkari what do you say?

**Comment by kristerkari** [comment]
I have a feeling that implementing the rule might be quite tricky. Matching the mixing code 1-1 with the code inside the file only works properly if you don't use variables or do math inside the mixing, so the [@mixin example in the description](https://github.com/kristerkari/stylelint-scss/issues/80#issue-186433758) would not really work.

Doing pre-parsing sounds fragile too. We would have to possibly parse quite many files and keep the parsed results in memory, so that could be really slow to do (we should try to keep our rules fast). For a big project it could just be too slow.

**Comment by branpar** [comment]
I've been thinking about how we might be able to improve this without the performance hit and I came up with an idea. Multiple new rules (specific for variables) under the `$-variable` category.  I'm thinking something like:

```
dollar-variable-after-color
dollar-variable-after-font
dollar-variable-after-border
...
```

As an example `dollar-variable-after-color` would check for the property `color `and enforce that only a variable is required there. This way it doesn't check to see if a variable already exists for the given hex/named color, just that we want to use a variable instead of a hex/color.

The other rule ideas I mentioned are to extend this to things like `font`, which would then allow enforcement of a variable to pass font family strings, and `border `and `border-color` for the `dollar-variable-after-border` option. 

Thoughts? I think this would be more universal and more appealing as you can enable it for specific aspects of the SCSS.  I haven't figured out a more performant way to enforce mixin usage yet.

**Comment by shawnbot** [comment]
FWIW, I’ve built something like this into our monolithic [stylelint-config-primer](/primer/stylelint-config-primer)’s new [`primer/variables` rule](https://github.com/primer/stylelint-config-primer/blob/master/plugins/README.md). I’m AFK right now so I can’t get too into it, but if anyone would be interested in helping me clean it up and abstract it out, I’d be more than happy to make that happen.
